lib/internal/mime.js

```diff
@@ -0,0 +1,279 @@
+  // console.dir({
+  //   pattern,
+  //   value,
+  //   _: [...value].map((c) => c.codePointAt(0)),
+  //   ret,
+  //   _1: (pattern.lastIndex = 0, NotHTTPTokenCodePoint.exec(value))
+  // });
+  return ret;
+};
+
+const encode = (value) => {
+  if (value.length === 0) return '""';
+  NotHTTPTokenCodePoint.lastIndex = 0;
+  const encode = test(NotHTTPTokenCodePoint, value);
+  if (!encode) return value;
+  const escaped = StringPrototypeReplace(value, /["\\]/g, '\\$&');
+  const ret = `"${escaped}"`;
+  return ret;
+};
+
+const MIMEStringify = (type, subtype, parameters) => {
+  let ret = `${type}/${subtype}`;
+  const entries = MIMEParamsData(parameters).entries();
+  let keyValuePair, done;
+  for (;;) {
+    ({ value: keyValuePair, done } =
+        MapIteratorPrototypeNext(entries));
+    if (done) break;
+    const [key, value] = keyValuePair;
+    const encoded = encode(value);
+    ret += `;${key}=${encoded}`;
+  }
+  return ret;
+};
+
+class MIMEParams {
+  #data;
+  constructor() {
+    this.#data = new Map();
+  }
+
+  delete(name) {
+    this.#data.delete(name);
+  }
+
+  get(name) {
+    const data = this.#data;
+    if (data.has(name)) {
+      return data.get(name);
+    }
+    return null;
+  }
+
+  has(name) {
+    return this.#data.has(name);
+  }
+
+  set(name, value) {
+    const data = this.#data;
+    NotHTTPTokenCodePoint.lastIndex = 0;
+    name = `${name}`;
+    value = `${value}`;
+    const invalidName = test(NotHTTPTokenCodePoint, name);
+    if (name.length === 0 || invalidName) {
+      throw new Error('Invalid MIME parameter name');
+    }
+    NotHTTPQuotedStringCodePoint.lastIndex = 0;
+    const invalidValue = test(NotHTTPQuotedStringCodePoint, value);
+    if (invalidValue) {
+      throw new Error('Invalid MIME parameter value');
+    }
+    data.set(name, value);
+  }
+
+  *entries() {
+    return yield* this.#data.entries();
+  }
+
+  *keys() {
+    return yield* this.#data.keys();
+  }
+
+  *values() {
+    return yield* this.#data.values();
+  }
+
+  *[SymbolIterator]() {
+    return yield* this.#data.entries();
+  }
+
+  static _data(o) {
+    return o.#data;
+  }
+}
+
+const MIMEParamsData = MIMEParams._data;
+delete MIMEParams._data;
+
+class MIME {
+  #type;
+  #subtype;
+  #parameters;
+  constructor(string) {
+    string = `${string}`;
+    const data = parse(string, { __proto__: null });
+    this.#type = data.type;
+    this.#subtype = data.subtype;
+    this.#parameters = new MIMEParams();
+    const paramsMap = MIMEParamsData(this.#parameters);
+    for (var i = 0; i < data.parameters.length; i++) {
+      const [k, v] = data.parameters[i];
+      paramsMap.set(k, v);
+    }
+  }
+
+  get type() {
+    return this.#type;
+  }
+
+  set type(v) {
+    v = `${v}`;
+    NotHTTPTokenCodePoint;
+    const invalidType = test(NotHTTPTokenCodePoint, v);
+    if (v.length === 0 || invalidType) {
+      throw new Error('Invalid MIME type');
+    }
+    this.#type = toASCIILower(v);
```

Comment:
How is ´toASCIILower` different from `toLowerCase` here?

Reply:
toLowerCase uses unicode default, not just ASCII. see [String.prototype.toLowerCase](https://tc39.es/ecma262/#sec-string.prototype.tolowercase):

> the elements are the result of toLowercase(cpList), according to the Unicode Default Case Conversion algorithm.

Reply:
Why is that relevant here?

Reply:
This is not the same outside of english:

[Per Unicode 12.1](http://www.unicode.org/versions/Unicode12.1.0/) some are even called out in section 3.13:

<img width="511" alt="Screen Shot 2019-11-28 at 8 08 47 AM" src="https://user-images.githubusercontent.com/234659/69812732-5d953100-11b6-11ea-9bc0-cc3e961f9888.png">